File: packages/android/vault/src/test/kotlin/app/keyweb/vault/MergeHistoryTest.kt
package app.keyweb.vault

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertContains
import kotlin.test.assertTrue

/**
 * What the undo window keeps when two devices edit at once.
 *
 * A single device archives what it overwrites. Two devices editing the same
 * field offline produce no such moment on either of them — each saw only its
 * own write — so the merge is the only place the losing value can be kept.
 * Parity with the web's `merge-history.test.ts`; these two implementations
 * have to agree, because the same vault passes through both.
 */
class MergeHistoryTest {

    private fun at(node: String, n: Int): Hlc =
        encodeHlc(HlcParts(wall = 1_700_000_000_000L + n, counter = 0, node = node))

    private fun past(item: ItemRecord, field: ItemField): List<String> =
        item.history.filter { it.field == field && it.value.isNotEmpty() }.map { it.value }

    private fun diverged(left: String, right: String): Pair<VaultState, VaultState> {
        val base = applyOps(
            emptyVault(),
            listOf(
                VaultOp.KeyringPut("k", at("a", 0), "ring", "Just mine"),
                VaultOp.ItemPut(
                    "i",
                    at("a", 1),
                    "bank",
                    "ring",
                    mapOf("title" to "Bank", "password" to "original"),
                ),
            ),
        )
        return applyOps(
            base,
            listOf(VaultOp.ItemPut("l", at("a", 5), "bank", "ring", mapOf("password" to left))),
        ) to applyOps(
            base,
            listOf(VaultOp.ItemPut("r", at("b", 9), "bank", "ring", mapOf("password" to right))),
        )
    }

    @Test
    fun `keeps the value that lost, not just the one that won`() {
        val (left, right) = diverged("from-the-laptop", "from-the-phone")
        val item = assertNotNull(mergeVaults(left, right).items["bank"])

        assertEquals("from-the-phone", item.field("password"))
        assertContains(past(item, "password"), "from-the-laptop")
        assertContains(past(item, "password"), "original")
    }

    @Test
    fun `says the same thing whichever way round the merge runs`() {
        val (left, right) = diverged("from-the-laptop", "from-the-phone")
        val one = assertNotNull(mergeVaults(left, right).items["bank"])
        val other = assertNotNull(mergeVaults(right, left).items["bank"])

        assertEquals(one.field("password"), other.field("password"))
        assertEquals(past(one, "password").sorted(), past(other, "password").sorted())
    }

    @Test
    fun `does not archive the loser twice when the merge is repeated`() {
        val (left, right) = diverged("from-the-laptop", "from-the-phone")
        val thrice = mergeVaults(mergeVaults(mergeVaults(left, right), right), left)
        val values = past(assertNotNull(thrice.items["bank"]), "password")

        assertEquals(1, values.count { it == "from-the-laptop" })
        assertEquals(1, values.count { it == "original" })
    }

    @Test
    fun `keeps a field's own history however busy another field is`() {
        var state = applyOps(
            emptyVault(),
            listOf(
                VaultOp.KeyringPut("k", at("a", 0), "ring", "Just mine"),
                VaultOp.ItemPut(
                    "i",
                    at("a", 1),
                    "bank",
                    "ring",
                    mapOf("title" to "Bank", "password" to "the-one-that-matters"),
                ),
                VaultOp.ItemPut("p", at("a", 2), "bank", "ring", mapOf("password" to "replaced-it")),
            ),
        )

        repeat(HISTORY_LIMIT * 3) { n ->
            state = applyOps(
                state,
                listOf(
                    VaultOp.ItemPut(
                        "n-$n",
                        at("a", 100 + n),
                        "bank",
                        "ring",
                        mapOf("note" to "note $n"),
                    ),
                ),
            )
        }

        val item = assertNotNull(state.items["bank"])
        assertContains(past(item, "password"), "the-one-that-matters")
        assertTrue(past(item, "note").size <= HISTORY_LIMIT)
    }

    @Test
    fun `caps each field on its own and keeps the newest of them`() {
        val entries = (0 until HISTORY_LIMIT + 4).map {
            HistoryEntry("password", "v$it", at("a", it))
        } + HistoryEntry("note", "a note", at("a", 0))

        val capped = capHistory(entries)
        assertEquals(HISTORY_LIMIT, capped.count { it.field == "password" })
        assertEquals(1, capped.count { it.field == "note" })
        assertEquals("v${HISTORY_LIMIT + 3}", capped.first().value)
        assertTrue(capped.none { it.value == "v0" })
    }

    private fun <T : Any> assertNotNull(value: T?): T = kotlin.test.assertNotNull(value)
}

File: packages/vault-core/tests/merge-history.test.ts
import { describe, expect, it } from "vitest";
import {
  applyOps,
  capHistory,
  emptyVault,
  encodeHlc,
  HISTORY_LIMIT,
  itemField,
  mergeVaults,
  pastValues,
} from "../src/index.js";

/**
 * What the undo window keeps when two devices edit at once.
 *
 * A single device archives what it overwrites: `item.put` pushes the value it
 * replaced into history. Two devices editing the same field offline produce no
 * such moment on either of them — each one only ever saw its own write — so
 * the merge is the only place the losing value can be kept. Without that, the
 * password one of them typed is gone from the item *and* from the list of what
 * it used to be, which is the one place somebody would go looking for it.
 */

/** What an item used to hold in one field, newest first. */
function past(item: Parameters<typeof pastValues>[0], field: string): string[] {
  return pastValues(item)
    .filter((entry) => entry.field === field)
    .map((entry) => entry.value);
}

function at(node: string, n: number) {
  return encodeHlc({ wall: 1_700_000_000_000 + n, counter: 0, node });
}

/** The same item, edited on two devices that have not spoken since. */
function diverged(leftValue: string, rightValue: string) {
  const base = applyOps(emptyVault(), [
    { kind: "keyring.put", opId: "k", ts: at("a", 0), keyringId: "ring", name: "Just mine" },
    {
      kind: "item.put",
      opId: "i",
      ts: at("a", 1),
      itemId: "bank",
      keyringId: "ring",
      fields: { title: "Bank", password: "original" },
    },
  ]);
  const left = applyOps(base, [
    {
      kind: "item.put",
      opId: "l",
      ts: at("a", 5),
      itemId: "bank",
      keyringId: "ring",
      fields: { password: leftValue },
    },
  ]);
  const right = applyOps(base, [
    {
      kind: "item.put",
      opId: "r",
      ts: at("b", 9),
      itemId: "bank",
      keyringId: "ring",
      fields: { password: rightValue },
    },
  ]);
  return { left, right };
}

describe("merging two devices that edited the same field", () => {
  it("keeps the value that lost, not just the one that won", () => {
    const { left, right } = diverged("from-the-laptop", "from-the-phone");
    const merged = mergeVaults(left, right);
    const item = merged.items["bank"]!;

    expect(itemField(item, "password")).toBe("from-the-phone");
    expect(past(item, "password")).toContain("from-the-laptop");
    // And the value from before either edit is still there.
    expect(past(item, "password")).toContain("original");
  });

  it("says the same thing whichever way round the merge runs", () => {
    const { left, right } = diverged("from-the-laptop", "from-the-phone");
    const one = mergeVaults(left, right).items["bank"]!;
    const other = mergeVaults(right, left).items["bank"]!;

    expect(itemField(one, "password")).toBe(itemField(other, "password"));
    expect(past(one, "password").sort()).toEqual(
      past(other, "password").sort(),
    );
  });

  it("does not archive the loser twice when the merge is repeated", () => {
    const { left, right } = diverged("from-the-laptop", "from-the-phone");
    const once = mergeVaults(left, right);
    const twice = mergeVaults(once, right);
    const thrice = mergeVaults(twice, left);

    const values = past(thrice.items["bank"]!, "password");
    expect(values.filter((value) => value === "from-the-laptop")).toHaveLength(1);
    expect(values.filter((value) => value === "original")).toHaveLength(1);
  });

  it("leaves history alone when both sides already agree", () => {
    const { left } = diverged("x", "y");
    const merged = mergeVaults(left, left);
    expect(past(merged.items["bank"]!, "password")).toEqual(
      past(left.items["bank"]!, "password"),
    );
  });
});

/**
 * The cap is per field, and that is the whole point.
 *
 * One cap across the item meant a run of edits to a note could push the only
 * archived password out of the window — the least valuable history evicting
 * the most valuable, silently, at the moment somebody needed it.
 */
describe("how much history is kept", () => {
  it("keeps a field's own history however busy another field is", () => {
    let state = applyOps(emptyVault(), [
      { kind: "keyring.put", opId: "k", ts: at("a", 0), keyringId: "ring", name: "Just mine" },
      {
        kind: "item.put",
        opId: "i",
        ts: at("a", 1),
        itemId: "bank",
        keyringId: "ring",
        fields: { title: "Bank", password: "the-one-that-matters" },
      },
      {
        kind: "item.put",
        opId: "p",
        ts: at("a", 2),
        itemId: "bank",
        keyringId: "ring",
        fields: { password: "replaced-it" },
      },
    ]);

    // Far more note edits than the cap, long after the password changed.
    for (let n = 0; n < HISTORY_LIMIT * 3; n += 1) {
      state = applyOps(state, [
        {
          kind: "item.put",
          opId: `n-${n}`,
          ts: at("a", 100 + n),
          itemId: "bank",
          keyringId: "ring",
          fields: { note: `note ${n}` },
        },
      ]);
    }

    const item = state.items["bank"]!;
    expect(past(item, "password")).toContain(
      "the-one-that-matters",
    );
    expect(past(item, "note").length).toBeLessThanOrEqual(HISTORY_LIMIT);
  });

  it("caps each field on its own and keeps the newest of them", () => {
    const entries = Array.from({ length: HISTORY_LIMIT + 4 }, (_, n) => ({
      field: "password",
      value: `v${n}`,
      ts: at("a", n),
    })).concat([{ field: "note", value: "a note", ts: at("a", 0) }]);

    const capped = capHistory(entries);
    expect(capped.filter((entry) => entry.field === "password")).toHaveLength(HISTORY_LIMIT);
    // The note survives a full password window rather than being crowded out.
    expect(capped.filter((entry) => entry.field === "note")).toHaveLength(1);
    // Newest first, and the oldest passwords are the ones dropped.
    expect(capped[0]!.value).toBe(`v${HISTORY_LIMIT + 3}`);
    expect(capped.map((entry) => entry.value)).not.toContain("v0");
  });
});
